File: python/src/aoc/year2015/day2.py
from aoc.util import load_input


def extract(line):
    """
    >>> extract('1x2x3')
    [1, 2, 3]
    >>> extract('3x2x1')
    [1, 2, 3]
    """
    return sorted(map(int, line.split("x")))


def calc_wrapping_paper(line):
    """
    >>> calc_wrapping_paper('2x3x4')
    58
    >>> calc_wrapping_paper('1x1x10')
    43
    """
    h, l, w = extract(line)
    return 3 * h * l + 2 * h * w + 2 * l * w


def calc_ribbon(line):
    """
    >>> calc_ribbon('2x3x4')
    34
    >>> calc_ribbon('1x1x10')
    14
    """
    h, l, w = extract(line)
    return 2 * (h + l) + h * l * w


def fun_and_sum(lines, fun):
    return sum(fun(line) for line in lines)


def part1(lines):
    return fun_and_sum(lines, calc_wrapping_paper)


def part2(lines):
    return fun_and_sum(lines, calc_ribbon)


if __name__ == "__main__":
    data = load_input(__file__, 2015, "2")
    print(part1(data))
    print(part2(data))

File: rust/src/aoc/year2015/day2.rs
pub fn yd() -> (usize, usize) {
    (2015, 2)
}

struct Dimensions {
    l: usize,
    w: usize,
    h: usize,
}

impl Dimensions {
    fn part1(&self) -> usize {
        3 * self.l * self.w + 2 * self.l * self.h + 2 * self.w * self.h
    }

    fn part2(&self) -> usize {
        2 * self.l + 2 * self.w + self.l * self.w * self.h
    }
}

fn solve<F>(data: &str, f: F) -> usize
where
    F: Fn(Dimensions) -> usize,
{
    let mut total = 0;
    let lines: Vec<&str> = data.split('\n').collect();
    for line in lines {
        if line.is_empty() {
            continue;
        }
        let mut n: Vec<usize> = line
            .split('x')
            .map(|s| s.parse::<usize>().unwrap())
            .collect();
        n.sort();
        let d = Dimensions {
            l: n[0],
            w: n[1],
            h: n[2],
        };
        total += f(d);
    }
    total
}

pub fn part1(data: &str) -> usize {
    solve(&data, |d| d.part1())
}

pub fn part2(data: &str) -> usize {
    solve(&data, |d| d.part2())
}
